flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/KNN.scala

```diff
@@ -0,0 +1,316 @@
+    override def fit(
+                      instance: KNN,
+                      fitParameters: ParameterMap,
+                      input: DataSet[T]): Unit = {
+      val resultParameters = instance.parameters ++ fitParameters
+
+      require(resultParameters.get(K).isDefined, "K is needed for calculation")
+
+      val blocks = resultParameters.get(Blocks).getOrElse(input.getParallelism)
+      val partitioner = FlinkMLTools.ModuloKeyPartitioner
+      val inputAsVector = input.asInstanceOf[DataSet[FlinkVector]]
+
+      instance.trainingSet = Some(FlinkMLTools.block(inputAsVector, blocks, Some(partitioner)))
+    }
+  }
+
+  /** [[PredictDataSetOperation]] which calculates k-nearest neighbors of the given testing data
+    * set.
+    * @tparam T Subtype of [[Vector]]
+    * @return The given testing data set with k-nearest neighbors
+    */
+
+  implicit def predictValues[T <: FlinkVector : ClassTag : TypeInformation] = {
+    new PredictDataSetOperation[KNN, T, (FlinkVector, Array[FlinkVector])] {
+      override def predictDataSet(
+                                   instance: KNN,
+                                   predictParameters: ParameterMap,
+                                   input: DataSet[T]):
+                                   DataSet[(FlinkVector, Array[FlinkVector])] = {
+        val resultParameters = instance.parameters ++ predictParameters
+
+        instance.trainingSet match {
+          case Some(trainingSet) =>
+            val k = resultParameters.get(K).get
+            val blocks = resultParameters.get(Blocks).getOrElse(input.getParallelism)
+            val metric = resultParameters.get(DistanceMetric).get
+            val partitioner = FlinkMLTools.ModuloKeyPartitioner
+
+            // attach unique id for each data
+            val inputWithId: DataSet[(Long, T)] = input.zipWithUniqueId
+
+            // split data into multiple blocks
+            val inputSplit = FlinkMLTools.block(inputWithId, blocks, Some(partitioner))
+
+            // join input and training set
+            val crossed = trainingSet.cross(inputSplit).mapPartition {
+              (iter, out: Collector[(FlinkVector, FlinkVector, Long, Double)]) => {
+                for ((training, testing) <- iter) {
+                  val queue = mutable.PriorityQueue[(FlinkVector, FlinkVector, Long, Double)]()(
+                    Ordering.by(_._4))
+
+                  // use a quadtree if (4^dim)Ntest*log(Ntrain)
+                  // < Ntest*Ntrain, and distance is Euclidean
+                  val useQuadTree = resultParameters.get(UseQuadTreeParam).getOrElse(
+                    training.values.head.size + math.log(math.log(training.values.length) /
+                      math.log(4.0)) < math.log(training.values.length) / math.log(4.0) &&
+                      (metric.isInstanceOf[EuclideanDistanceMetric] ||
+                        metric.isInstanceOf[SquaredEuclideanDistanceMetric]))
```

Comment:
If the user set `UseQuadTreeParam` to true and use `CosineDistanceMetric` as distance metric, the value of `useQuadTree` will be true. It can occur unexpected error.

What is exact meaning of `UseQuadTreeParam`? Does it mean force-use of quad tree?